### PR TITLE
Feature bandwidth burst

### PR DESCRIPTION
**Pull request by dotSlashLu**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Previously CNI cniBandwidthEntry.{In,E}gressBurst is not used and
defaults to the MaxInt32, this renders bandwidth control nearly
useless for cost control since the peak bandwidth is basically
unpredictable. See https://stackoverflow.com/questions/59729354/set-burst-for-bandwidth-limit-for-a-pod.

This patch introduces two new annotations
kubernetes.io/ingress-busrt and kubernetes.io/egress-burst to setup
cniBandwidthEntry.IngressBurst and cniBandwidthEntry.EgressBurst.

**Which issue(s) this PR fixes**:
N/A

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Introduces two new annotations
kubernetes.io/ingress-busrt and kubernetes.io/egress-burst to setup
cniBandwidthEntry.IngressBurst and cniBandwidthEntry.EgressBurst
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

### CODE REVIEW

**Comment by dotSlashLu** [comment]
I signed it

File: pkg/kubelet/dockershim/network/cni/cni.go [hunk 1]
```diff
@@ -435,12 +433,20 @@ func (plugin *cniNetworkPlugin) buildCNIRuntimeConf(podName string, podNs string
...
+			} else {
+				bandwidthParam.IngressBurst = math.MaxInt32
+			}
 		}
 		if egress != nil {
-			bandwidthParam.EgressRate = int(egress.Value())
-			bandwidthParam.EgressBurst = math.MaxInt32 // no limit
+			bandwidthParam.EgressRate = int(egress.Rate.Value())
+			if egress.Burst != nil {
+				bandwidthParam.EgressBurst = int(egress.Burst.Value())
```
Comment (aojea) [inline pkg/kubelet/dockershim/network/cni/cni.go]: I'm not good with this type conversions, but my understanding is that [`int` may be 32 or 64 bits](https://stackoverflow.com/questions/21491488/what-is-the-difference-between-int-and-int64-in-go).
Should we cap it to `math.MaxInt32`?

Comment (dotSlashLu) [reply pkg/kubelet/dockershim/network/cni/cni.go]: Hi aojea, thanks for you advice. I followed previous implementation on assigning int64 rate value to `cniBandwidthEntry` field of int, see deleted line 442 of this file. 

And I think you're right about `int`, see: https://golang.org/pkg/builtin/#int, it's platform specific. I also did some experiment on it, actually `int(int64)` will remain `int64`'s size and will not be truncated on 64-bit platform, but on 32-bit platform(like the playground) it will result an error: https://play.golang.org/p/c34Z2QkPfKY.

If supporting 32-bit platform is not a goal, and considering rate and burst are in bit, they could be quit large, I don't think capping it to `MaxInt32`(just around 2147 Mbit) is a good idea.

File: pkg/util/bandwidth/utils.go [hunk 2]
```diff
@@ -35,32 +41,69 @@ func validateBandwidthIsReasonable(rsrc *resource.Quantity) error {
...
+		}
+		burstValue, err := resource.ParseQuantity(str)
 		if err != nil {
-			return nil, nil, err
+			return nil, err
 		}
-		egress = &egressValue
-		if err := validateBandwidthIsReasonable(egress); err != nil {
-			return nil, nil, err
+		if err := validateBandwidthIsReasonable(&burstValue); err != nil {
```
Comment (aojea) [inline pkg/util/bandwidth/utils.go]: does bandwidth validation apply to burst validation?
reading the bandwidth plugin docs 
 https://github.com/containernetworking/plugins/blob/master/plugins/meta/bandwidth/README.md
burst are bits and bandwidth bps

File: pkg/kubelet/dockershim/network/cni/cni.go (hunk 1 above)
Comment (aojea) [reply pkg/kubelet/dockershim/network/cni/cni.go]: thanks for the explanation, it makes sense

File: pkg/util/bandwidth/utils_test.go [hunk 3]
```diff
@@ -30,41 +30,72 @@ func TestExtractPodBandwidthResources(t *testing.T) {
...
+		expectedEgress  *BandwidthLimit
 		expectError     bool
 	}{
 		{
 			pod: &api.Pod{},
 		},
 		{
-			pod:             testPod("10M", ""),
-			expectedIngress: &ten,
+			pod: testPod("10M", "", "10M", ""),
```
Comment (aojea) [inline pkg/util/bandwidth/utils_test.go]: you should add test cases with for `Burst`, happy cases, check that it errors if burst is set but rate not, ...

File: pkg/util/bandwidth/utils.go (hunk 2 above)
Comment (dotSlashLu) [reply pkg/util/bandwidth/utils.go]: Though bandwidth is in bps the units in code are actually both represented as bits since 1 bit per second just equals 1 bit in value, I don't think this differences these two values. 

But I think if we have to do upper bound validation in this step, we have to know what range is rational for a burst value. I think the upper bound of burst shares with bandwidth's, since rationally they all should be the bandwidth of the network interface card. So I think there're no problems sharing this validation.

Actually, I don't think the validation for upper bound is necessary. The current implementation of the bandwidth plugin just feeds this two values to the `tc` command, which doesn't have a limit for these parameters. And the current limit of 1P is fairly large, network interface card bandwidths are just around ten thousand Mbit range after all.

What's your opinion?

File: pkg/util/bandwidth/utils_test.go (hunk 3 above)
Comment (dotSlashLu) [reply pkg/util/bandwidth/utils_test.go]: done

Comment (aojea) [reply pkg/util/bandwidth/utils_test.go]: what about?
testPod("10M", "10M", "10M", "10M"),
testPod("10M", "10M", "10M", ""),
testPod("10M", "", "10M", "10M"),
testPod(lower than minBW, "10M", "10M", "10M") must err
testPod(greater than maxBW, "10M", "10M", "10M") must err
testPod("10M", lower than min burst, "10M", "10M") must err
testPod("10M", greater than max burst, "10M", "10M") must err

File: pkg/util/bandwidth/utils.go (hunk 2 above)
Comment (aojea) [reply pkg/util/bandwidth/utils.go]: I personally think we must mimic the `bandwidth` plugin validation and using a name like `validateBandwidthAndBurstIsReasonable`

https://github.com/containernetworking/plugins/blob/cf187287af3f50dfd2de6bea4fec62c1373caf3c/plugins/meta/bandwidth/main.go#L105-L118

File: pkg/kubelet/dockershim/network/cni/cni.go (hunk 1 above)
Comment (aojea) [reply pkg/kubelet/dockershim/network/cni/cni.go]: looking at the CNI bandwidth plugin it seems that `burst/8 >= math.MaxUint32:` is not possible

https://github.com/containernetworking/plugins/blob/f5c3d1b1bab6ee72a32702c22ebb0f008580ef2d/plugins/meta/bandwidth/main.go#L113-L114

File: pkg/util/bandwidth/utils_test.go (hunk 3 above)
Comment (dotSlashLu) [reply pkg/util/bandwidth/utils_test.go]: that's precise and thorough enough, thanks for the advice, added some comment and updated, please review again.

File: pkg/util/bandwidth/utils.go (hunk 2 above)
Comment (dotSlashLu) [reply pkg/util/bandwidth/utils.go]: good catch, thanks.

File: pkg/util/bandwidth/utils.go [hunk 4]
```diff
@@ -17,50 +17,109 @@ limitations under the License.
...
 }
 
-// ExtractPodBandwidthResources extracts the ingress and egress from the given pod annotations
-func ExtractPodBandwidthResources(podAnnotations map[string]string) (ingress, egress *resource.Quantity, err error) {
+func validateBurstIsReasonable(rsrc *resource.Quantity) error {
+	if rsrc.Value() < minBurst.Value() {
+		return fmt.Errorf("resource is unreasonably small (< 1kbit)")
+	}
+	if rsrc.Value() > maxBurst.Value() {
+		return fmt.Errorf("resoruce is unreasonably large (> 4GB)")
```
Comment (aojea) [inline pkg/util/bandwidth/utils.go]: nit, you carried on the typo :)

Comment (dotSlashLu) [reply pkg/util/bandwidth/utils.go]: haha, sorry about that, I've updated the error message to clarify the source type, please see the latest commit.

**Comment by aojea** [comment]
LGTM
we should squash the commits before merging
/assign @dcbw , can you please take a look Dan?

**Comment by thockin** [comment]
Annotations make a poor API.

The bandwidth-related annotations have existed since WAY back as a sort of proof of concept.  We carried them forward for compatibility, but they have always been optional and somewhat under-specified.  Nobody is really happy about them.  I do not think we should add more annotations, especially without giving due consideration to whether and how we could do this "properly".

TL;DR: Where's the KEP?

**Comment by aojea** [comment]
> Annotations make a poor API.
> 
> The bandwidth-related annotations have existed since WAY back as a sort of proof of concept. We carried them forward for compatibility, but they have always been optional and somewhat under-specified. Nobody is really happy about them. I do not think we should add more annotations, especially without giving due consideration to whether and how we could do this "properly".
> 
> TL;DR: Where's the KEP?

/lgtm cancel

sorry, I didn't have that context and adding it seemed correct because of the code comments
> // NOTE: it's not used for now and default to 0.
 the functionality seems correctly implemented

**Comment by dotSlashLu** [comment]
Thanks for the suggestion, I'll add a KEP. If annotations are not ideal, could you please direct me a better way to make such configurations?

**Comment by thockin** [comment]
That's what the KEP is for.  One option. Is to just do more annotations.
Another option is to use real fields to configure pod networking.

We need to consider implement ability in different environments, plugins,
etc.

**Comment by thockin** [comment]
I'm going to close this, for now, when you have a kep ready, assign to me and we can discuss
